### PR TITLE
Add guard for dependencies recursively depending on kubernetes or staging modules

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind cleanup
/sig architecture
/area code-organization

#### What this PR does / why we need it:

Protects against dependency cycles. I thought we already had this, but apparently only checked for things depending back on k8s.io/kubernetes

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: hack/lint-dependencies.sh [hunk 1]
```diff
@@ -47,6 +47,19 @@ rc=0
 # The array and map in `unwanted-dependencies.json` are in alphabetical order.
 go run k8s.io/kubernetes/cmd/dependencyverifier "${KUBE_ROOT}/hack/unwanted-dependencies.json"
 
+k8s_module_regex="k8s[.]io/(kubernetes"
```
Comment (BenTheElder) [inline hack/lint-dependencies.sh:50]: we bothered to escape the `.` here but we aren't for the repos below, should we fix that?

**Review by BenTheElder — COMMENTED** [review]
generally LGTM

File: hack/lint-dependencies.sh (hunk 1 above)
Comment (liggitt) [reply hack/lint-dependencies.sh:50]: which repos below? list_staging_repos only gives the last path segment, none of which contain `.`

**Pull request merged**